File: haskell/collatz-conjecture/src/CollatzConjecture.hs
module CollatzConjecture (collatz) where

collatz :: Integer -> Maybe Integer
collatz x
  | x <= 0     = Nothing
  | otherwise  = Just (step x 0)
  where
    step :: Integer -> Integer -> Integer
    step n acc
      | n == 1    = acc
      | even n    = step (div n 2) (acc + 1)
      | otherwise = step (3 * n + 1) (acc + 1)

File: erlang/collatz-conjecture/src/collatz_conjecture.erl
-module(collatz_conjecture).

-export([steps/1]).

steps(N) when N =< 0 -> error(badarg);
steps(N) -> step(N, 0).

step(1, Acc) -> Acc;
step(N, Acc) when N rem 2 == 0 -> step(N div 2, Acc + 1);
step(N, Acc) -> step(3 * N + 1, Acc + 1).
